chips/sam4l/src/usbc/common_register.rs

```diff
@@ -0,0 +1,244 @@
+#![allow(dead_code)]
+
+use core::marker::PhantomData;
+
+/// A memory-mapped register
+#[derive(Copy, Clone)]
+pub struct Reg<T> {
+    addr: *mut u32,
+    phantom: PhantomData<*mut T>,
+}
+
+impl<T: FromWord + ToWord> Reg<T> {
+    pub const fn new(addr: *mut u32) -> Self {
+        Reg { addr: addr, phantom: PhantomData }
+    }
+
+    #[inline]
+    pub fn read(self) -> T {
+        T::from_word(self.read_word())
+    }
+
+    #[inline]
+    pub fn write(self, val: T) {
+        unsafe { ::core::ptr::write_volatile(self.addr, T::to_word(val)); }
+    }
+
+    #[inline]
+    pub fn read_word(self) -> u32 {
+        unsafe { ::core::ptr::read_volatile(self.addr) }
+    }
+}
+
+impl Reg<u32> {
+    #[inline]
+    pub fn set_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w | bit);
+    }
+
+    #[inline]
+    pub fn clr_bit(self, bit_index: u32) {
+        let w = self.read();
+        let bit = 1 << bit_index;
+        self.write(w & (!bit));
+    }
+}
+
+/// A write-only memory-mapped register, where any zero bits written have no effect
```

Comment:
I don't think "where any zero bits written have no effect" is important here. The type seems equally functional for write-only registers where 0-bits have an effect and for those that don't.